Bitcoin for SWAP & HTLC (#35)
comit-network/RFCs: RFC-005-SWAP-Bitcoin-basic.md

# Bitcoin Basic HTLC Atomic Swap

- RFC-Number: 005
- Status: Draft
- Discussion-issue: -
- Created on: 01 Mar. 2019

# Table of contents

<!-- toc -->

- [Description](#description)
- [Motivation](#motivation)
- [Content](#content)
  - [Identities](#identities)
  - [Scripts](#scripts)
    - [Variables](#variables)
    - [HTLC](#htlc)
      - [Human friendly - Opcode representation](#human-friendly---opcode-representation)
      - [Implementation friendly - Hex representation](#implementation-friendly---hex-representation)
    - [Redeem](#redeem)
    - [Refund](#refund)
    - [Timeouts](#timeouts)
}- [Registry extension](#registry-extension)

<!-- tocstop -->

## Description

> This RFC defines the Bitcoin ledger and the Bitcoin asset for use in an execution [RFC002](./RFC-002-SWAP.md) SWAP with `comit-rfc-003` ([RFC003](./RFC-003-SWAP-basic.md)) as the SWAP protocol.

Reviewer: :wrench: At first read, this sounds like you are repeating yourself from RFC4. Maybe start the sentence with what it actually defines?

```suggestion
This RFC defines how to execute an [RFC002](./RFC-002-SWAP.md) SWAP with `comit-rfc-003` ([RFC003](./RFC-003-SWAP-basic.md)) as the SWAP protocol for the Bitcoin ledger and the Bitcoin asset.
```

Reviewer 2: Yeah I originally had this exact wording but I changed it because this RFC doesn't exclusively determine how to execute an RFC3 swap (rather both ledger's RFCs do). I will think some more about how to word this better.

For definitions of the Bitcoin ledger and asset see [RFC004](./RFC-004-SWAP-Bitcoin).

To fulfil the requirements of RFC003 this RFC defines:

- The *identity* to be used when negotiating a SWAP on the Bitcoin ledger.
- How to construct a Hash Time Lock Contract (HTLC) to lock the Bitcoin asset on the Bitcoin blockchain.
- How to deploy, redeem and refund the HTLC during the execution phase of RFC003

> ## Motivation

Reviewer 2: Done. yeah it was a bit of a stretch.

| Ledger   | Identity Name | JSON Encoding            | Description                                                                                  |
|:----     |:-------       |:-------------            | -------------------------------------------------------------------------------------------- |
| Bitcoin  | `pubkeyhash`  | `hex-encoded-bytes (20)` | The result of applying SHA256 and then RIPEMD160 to a user's SECP256k1 compressed public key |

## Hash Time Lock Contract

### Hash Functions

The `hash_function` header must be set to one of the following values if Bitcoin is used as a ledger:

- SHA-256

### Parameters

The parameters for the Bitcoin HTLC follow [RFC003](./RFC-003-SWAP-basic.md) and are described concretely in the following table:

| Variable        | Description                                                                                                               |
|:----------------|:--------------------------------------------------------------------------------------------------------------------------|
| asset           | The quantity of satoshi                                                                                                   |

> | secret          | The secret (32 bytes for SHA-256)                                                                                         |

Reviewer 3: :question: Could it be described without using the word "secret"?

Reviewer 2: Err I don't even know what it's doing there. Deleted!

| secret_hash     | The SHA-256 hash of the `secret` (32 bytes for SHA-256)                                                                   |
| redeem_identity | The `pubkeyhash` of the redeeming parry                                                                                   |

> | refund_identity | The `pubkeyhash` of the refunding party                                                                                   |

Reviewer 3: :upside_down_face:
```suggestion
| refund_identity | The `pubkeyhash` of the funding party                                                                                   |
```

| `8876a9`          | 40                     | 42                    | 3      | `OP_EQUALVERIFY OP_DUP OP_HASH160`                                         |
| `redeem_identity` | 43                     | 74                    | 32     | See [Parameters](#parameters)                                              |
| `67`              | 75                     | 75                    | 1      | `OP_ELSE`                                                                  |
| `timestamp`       | 76                     | 79                    | 4      | See [Parameters](#parameters)                                              |
| `b17576a9`        | 80                     | 83                    | 4      | `OP_CHECKLOCKTIMEVERIFY OP_DROP OP_DUP OP_HASH160`                         |
| `refund_identity` | 84                     | 115                   | 32     | See [Parameters](#parameters)                                              |
| `6888ac`          | 116                    | 118                   | 3      | `OP_ENDIF OP_EQUALVERIFY OP_CHECKSIG`                                      |

## Execution Phase

### Deployment

> The party who is required to deploy the Bitcoin HTLC (the refunder) compiles the contract as described in the previous section.

Reviewer 3: I understand why you're not using Alice and Bob, because it would be confusing in the larger context. But I would use "funder"/"deployer" instead of "refunder", since it's the action that will happen most often and puts emphasis on the successful outcome of RFC003.

```suggestion
The party who is required to deploy the Bitcoin HTLC (the funder) compiles the contract as described in the previous section.
```

> Then, they MUST send a transaction to the Bitcoin Blockchain with an output with value exactly equal to the quantity of satoshi specified by `asset` and a Pay-To-Witness-Script-Hash (P2WSH) `scriptPubKey` derived from the contract. See [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#specification)

Reviewer 3: :question: Is this really a `MUST`? You don't use this sort of language in the rest of this section anyway, but to me this is a `MAY` unless you want to have an opinion on whether the "funder"/"deployer"/"refunder" should proceed with the protocol. What if Bitcoin is `beta_ledger` and Alice took too long to fund `alpha_asset`?

Reviewer 2: I changed it :)

| `01`             | Used to activate the redeem path in the `OP_IF`                                                         |
| contract         | The compiled contract (as generally required when redeeming from a P2WSH output)                        |

For how to use this to construct the redeem transaction see [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#transaction-id).

#### Refund

To refund the HTLC, the refunder should submit a transaction to the blockchain which spends the P2WSH output.
The refunder can use the following witness data to spend the output after the `expiry`:

| Data             | Description                                                                                             |
|:-----------------|:--------------------------------------------------------------------------------------------------------|
| refund_signature | A valid SECP256k1 ECDSA DER encoded signature on the transaction with respect to the `refund_pubkey`    |
| redeem_pubkey    | The 33 byte SECP256k1 compressed public key that was hashed to produce the pubkeyhash `redeem_identity` |

> | `00`              | Used to activate the refund path in the `OP_IF`                                                         |

Reviewer 3: :x: The table is a little broken :construction_worker_man: 
```suggestion
| `00`             | Used to activate the refund path in the `OP_IF`                                                         |
```

Reviewer 2: Why would deleting a space help???👷

Reviewer 3: I'm no architect :building_construction: but it lines up all the pipes.

| contract         | The compiled contract (as generally required when redeeming from a P2WSH output)                        |

## Registry extension

This RFC extends the [registry](./registry.md) in the following ways:

- **pubkeyhash**: The identities table now includes:

> | Ledger   | Identity Name | JSON Encoding            | Description                                                                                  |

Reviewer 3: :x: This table is broken, but I don't know how to fix it :disappointed:

|:----     |:-------       |:-------------            | -------------------------------------------------------------------------------------------- |
| Bitcoin  | `pubkeyhash`  | `hex-encoded-bytes (20)` | The result of applying SHA256 and then RIPEMD160 to a user's SECP256k1 compressed public key |

## Hash Time Lock Contract

### Hash Functions

SHA-256 is the only value the `hash_function` header may take if Bitcoin is used as a ledger.

### Parameters

The parameters for the Bitcoin HTLC follow [RFC003](./RFC-003-SWAP-basic.md) and are described concretely in the following table:

| Variable        | Description                                                                 |
|:----------------|:----------------------------------------------------------------------------|
| asset           | The quantity of satoshi                                                     |
| secret_hash     | The hash of the `secret` (32 bytes for SHA-256)                             |
| redeem_identity | The `pubkeyhash` requ the redeeming party                                   |
| refund_identity | The `pubkeyhash` of the refunding party                                     |

> | expiry          | The absolute UNIX timestamp in seconds after which the HTLC can be refunded |

Reviewer 2: I would like to add a note that implementations MUST check that this is a valid time to prevent against an attack where someone sends a very low value for expiry which looks is actually a absolute bock height rather than an unix epoch.

Reviewer 2: I've added a note about this where we talk about `OP_CHECKLOCKTIMEVERIFY`.

| contract         | The compiled contract (as generally required when redeeming from a P2WSH output)                        |

To be notified of the refund event both parties may watch the blockchain for transactions that spend from the output and check that the witness data is in the above form.

## Registry extension

This RFC extends the [registry](./registry.md) with an identity definition for the Bitcoin ledger:

| Ledger  | Identity Name | JSON Encoding            | Description                                                                           |
|:--------|:--------------|:-------------------------|---------------------------------------------------------------------------------------|
| Bitcoin | `pubkeyhash`  | `hex-encoded-bytes (20)` | The result of applying SHA256 and then RIPEMD160 to a SECP256k1 compressed public key |

# Examples

## Example RFC003 message body

This following shows what a SWAP message looks like where the `alpha_ledger` is Bitcoin, the `alpha_asset` is 1 Bitcoin (with `...` being used where the value is only relevant for the `beta_ledger`).

> ``` json

Reviewer 4: Can we have a complete example here, i.e. without `{...}`?

Reviewer: I am not sure if I want that. It would add unnecessary complexity and detail that this RFC is not concerned with. Like, which ledger to choose. Technically, no other ledgers are defined yet. We just added them to the registry.

To me, the example is clear enough without those details.

Reviewer 4: In that particular case it's about 5 additional lines and not more complex :) 
I'm personally always annoyed when finding incomplete examples.

Reviewer 2: I find incomplete examples in *software libraries* annoying. But I think this example wouldn't benefit from being complete as it is just trying to demonstrate how to compile the HTLC. 

There is a complete example in RFC-003. I would be fine with deleting this one in favour of that one (and just having the table and resulting contract and addresses.

| `secret_hash`     | 7                      | 39                    | 32     | See [Parameters](#parameters)                                              |
| `8876a9`          | 40                     | 42                    | 3      | `OP_EQUALVERIFY OP_DUP OP_HASH160`                                         |
| `redeem_identity` | 43                     | 64                    | 20     | See [Parameters](#parameters)                                              |
| `67`              | 65                     | 65                    | 1      | `OP_ELSE`                                                                  |
| `expiry`          | 66                     | 69                    | 4      | See [Parameters](#parameters)                                              |
| `b17576a9`        | 70                     | 73                    | 4      | `OP_CHECKLOCKTIMEVERIFY OP_DROP OP_DUP OP_HASH160`                         |
| `refund_identity` | 74                     | 93                    | 20     | See [Parameters](#parameters)                                              |

> | `6888ac`          | 94                     | 96                    | 3      | `OP_ENDIF OP_EQUALVERIFY OP_CHECKSIG`                                      |

Reviewer 2: Can I get someone to go through this and verify it's correct?

# Bitcoin Basic HTLC Atomic Swap

- RFC-Number: 005
- Status: Draft
- Discussion-issue: -
- Created on: 01 Mar. 2019

# Table of Contents
<!-- markdown-toc start -->

- [Description](#description)

> - [Motivation](#motivation)

Author: ```suggestion
```

> - [Bitcoin Identity](#bitcoin-identity)

Author: Update your toc.

- The [identity](./RFC-003-SWAP-basic.md#identity) to be used when negotiating a SWAP on the Bitcoin ledger.
- How to construct a Hash Time Lock Contract (HTLC) to lock the Bitcoin asset on the Bitcoin blockchain.
- How to deploy, redeem and refund the HTLC during the execution phase of [RFC003](./RFC-003-SWAP-basic.md).

## The Bitcoin Identity

The Identity to be used on Bitcoin is the 20-byte *pubkeyhash* which is the result of applying SHA-256 and then RIPEMD-160 to a user's compressed SECP256k1 public key.
The compressed public key is used because it needs to be compatible with segwit transactions (see [BIP143](https://github.com/bitcoin/bips/blob/master/bip-0143.mediawiki#Restrictions_on_public_key_type)).

While it may seem more intuitive to use a Bitcoin *address* as the identity, the pubkeyhash better fits the definition of identity given in [RFC003](./RFC-003-SWAP-basic.md).
Using a Bitcoin address as the identity would require implementations to do a number of cumbersome validation steps such as verifying that it is a p2pkh or p2wpkh address, extracting the pubkeyhash and validating the network.

> In the JSON encoding, A *pubkeyhash* MUST be encoded as a 20 byte hex string.

Author: ```suggestion
In the JSON encoding, A *pubkeyhash* MUST be encoded as a 20-byte hex string.
```
Stick to one way of spelling it

OP_ELSE
    <expiry> OP_CHECKLOCKTIMEVERIFY OP_DROP
    OP_DUP OP_HASH160 <refund_identity>
OP_ENDIF
OP_EQUALVERIFY
OP_CHECKSIG
```

As required by [RFC003](./RFC-003-SWAP-basic.md), this HTLC uses absolute time locks to check whether `expiry` has been reached.
Specifically, `OP_CHECKLOCKTIMEVERIFY` (see [BIP65](https://github.com/bitcoin/bips/blob/master/bip-0065.mediawiki)) is used to compare the time in the block's header to the `expiry` in the contract.

Implementations MUST consider an `expiry` value below `500000000` for a Bitcoin HTLC to be invalid due to the [`lock_time`](https://en.bitcoin.it/wiki/Protocol_documentation#tx) transaction field interpreting values below `500000000` as block heights.

To compute the exact Bitcoin script bytes of the contract, implementations should use the following offset table:

| Data              | Position of first byte | Position of last byte | Length (bytes) | Description                                                                |
|:------------------|:-----------------------|:----------------------|:---------------|:---------------------------------------------------------------------------|

> | `6382012088a820`  | 0                      | 6                     | 7              | `OP_IF OP_SIZE 32 OP_EQUALVERIFY OP_SHA256` + `PUSH32` for the secret_hash |

Author: ❓ should we use a consistent approach on how we write this table? I am asking because you put less info in the Ethereum one.

Reviewer 2: I'm not sure. I originally started going into full detail on the Ethereum one too but it was too painful. Consistency isn't the most important attribute in RFCs and is certainly behind readability so I'd say we leave it like this for now. If we find it would be easier to implement if we just had blob of hex then we can do that.

| `redeem_identity` | 43                     | 64                    | 20             | See [Parameters](#parameters)                                              |
| `67`              | 65                     | 65                    | 1              | `OP_ELSE`                                                                  |
| `expiry`          | 66                     | 69                    | 4              | See [Parameters](#parameters)                                              |
| `b17576a9`        | 70                     | 73                    | 4              | `OP_CHECKLOCKTIMEVERIFY OP_DROP OP_DUP OP_HASH160`                         |
| `refund_identity` | 74                     | 93                    | 20             | See [Parameters](#parameters)                                              |
| `6888ac`          | 94                     | 96                    | 3              | `OP_ENDIF OP_EQUALVERIFY OP_CHECKSIG`                                      |

## Execution Phase

The following section describes how both parties should interact with the Bitcoin blockchain during the [RFC003 execution phase](./RFC-003-SWAP-basic.md#execution-phase).

### Deployment

> At the start of the deployment stage, both parties compile the contract into bytes as described in the previous section.

Author: ```suggestion
At the start of the deployment stage, both parties construct the contract into bytes as described in the previous section.
```
See https://github.com/comit-network/RFCs/pull/38#discussion_r265379886

Moreover, is it really `bytes` as it is always kept in hex representation? Only bitcoind does the translation hex -> byte.

Reviewer 2: I think compiles is the right word because we're compiling a contract from a contract template + parameters. I removed the the thing about "into bytes", because as you say we don't want to imply a particular representation it should compile into.

|:-----------------|:--------------------------------------------------------------------------------------------------------|
| redeem_signature | A valid SECP256k1 ECDSA DER encoded signature on the transaction with respect to the `redeem_pubkey`    |
| redeem_pubkey    | The 33 byte SECP256k1 compressed public key that was hashed to produce the pubkeyhash `redeem_identity` |
| secret           | The pre-image of the `secret_hash` under the `hash_function`                                            |
| `01`             | A single byte used to activate the redeem path in the `OP_IF`                                           |
| contract_script  | The compiled contract (as generally required when redeeming from a P2WSH output)                        |

For how to use this witness data to construct the redeem transaction see [BIP141](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki#transaction-id).

To be notified of the redeem event, both parties MAY watch the blockchain for transactions that spend from the output and check that the witness data is in the above form.

> If Bitcoin is the `beta_ledger`, then the funder (Bob) MUST watch for such a transaction and then extract the `secret` from the witness data and continue the protocol.

Author: ```suggestion
If Bitcoin is the `beta_ledger`, then the *funder* (Bob) MUST watch for such a transaction and then extract the `secret` from the witness data and continue the protocol.
```
To be consistent

Reviewer 2: I think you should only put italics on the first use.

> If Bitcoin is the `beta_ledger`, then the funder (Bob) MUST watch for such a transaction and then extract the `secret` from the witness data and continue the protocol.

Reviewer 3: :upside_down_face: Same as for Ethereum, this is the first time that `beta_ledger` and Bob are mentioned in this document. I would link to the [relevant section](https://github.com/comit-network/RFCs/blob/master/RFC-003-SWAP-basic.md#4-bob-redeems) in RFC003. Could also extract it into its own subsection.

Reviewer 2: I linked to it but it sucks because you can't really figure out what is meant from it by looking at RFC003. I made a comment here: https://github.com/comit-network/RFCs/issues/19

### Refund

> To refund the HTLC, the funder MUST submit a transaction to the blockchain which spends the P2WSH output.

Author: ```suggestion
To refund the HTLC, the *funder* MUST submit a transaction to the blockchain which spends the P2WSH output.
```

Reviewer 2: It's in italics in the deploy section already.

| refund_pubkey    | The 33 byte SECP256k1 compressed public key that was hashed to produce the pubkeyhash `refund_identity` |
| `00`             | A single byte used to activate the refund path in the `OP_IF`                                           |
| contract_script  | The compiled contract (as generally required when redeeming from a P2WSH output)                        |

To be notified of the refund event, both parties MAY watch the blockchain for transactions that spend from the output and check that the witness data is in the above form.

## Registry extension

This RFC extends the [registry](./registry.md#identities) with an identity definition for the Bitcoin ledger:

| Ledger  | Identity Name | JSON Encoding            | Description                                                                           |
|:--------|:--------------|:-------------------------|---------------------------------------------------------------------------------------|
| Bitcoin | `pubkeyhash`  | `hex-encoded-bytes (20)` | The result of applying SHA-256 and then RIPEMD-160 to a SECP256k1 compressed public key |

> # Examples

Reviewer: :wrench: Could be explicit that it contains test vectors.

```suggestion
# Examples and test vectors
```

Reviewer 2: Changed to Examples/Test vectors

## RFC003 SWAP REQUEST

The following shows an [RFC003](RFC-003-SWAP-basic.md) SWAP REQUEST where the `alpha_ledger` is Bitcoin, the `alpha_asset` is 1 Bitcoin (with `...` being used where the value is only relevant for the `beta_ledger`).

``` json
{
  "type": "SWAP",
  "headers": {
    "alpha_ledger": {
      "value": "bitcoin",
      "parameters": { "network": "mainnet" }
    },
    "beta_ledger": {...},
    "alpha_asset": {
      "value": "bitcoin",
      "parameters": { "quantity": "100000000" }
    },
    "beta_asset": {...},
    "protocol": {
        "value" : "comit-rfc-003",
        "parameters" : { "hash_function" : "SHA-256" }
    }
  },
  "body": {
    "aplha_ledger_refund_identity": "1925a274ac004373bb5429553bdb55c40e57b124",
    "alpha_expiry": 1552263040,
    "secret_hash" : "51a488e06e9c69c555b8ad5e2c4629bb3135b96accd1f23451af75e06d3aee9c",
    "beta_ledger_redeem_identity" : "...",
    "beta_expiry" : ...
  },
}
```

> Note, the secret for the `secret_hash` is `51a488e06e9c69c555b8ad5e2c4629bb3135b96accd1f23451af75e06d3aee9c`.

Reviewer: :x: How can this be true if both values are identical?

Reviewer 2: Well spotted!

    "secret_hash" : "1f69c8745f712da03fdd43486ef705fc24f3e34d54cf44d967cf5cd4204c835e",
    "beta_ledger_redeem_identity" : "...",
    "beta_expiry" : ...
  },
}
```

Note, the secret for the `secret_hash` is `51a488e06e9c69c555b8ad5e2c4629bb3135b96accd1f23451af75e06d3aee9c`.

## RFC003 SWAP RESPONSE
A valid `RESPONSE` to the above `REQUEST` could look like:

``` json
{
  "status" : "OK00",
  "body": {
     "alpha_ledger_redeem_identity": "c021f17be99c6adfbcba5d38ee0d292c0399d2f5",
     "beta_ledger_refund_identity": "..."
  }
}
```

## HTLC

The above `REQUEST` and `RESPONSE` results in the following parameters to the HTLC:

| Parameter       | value                                                              |
|:----------------|--------------------------------------------------------------------|
| redeem_identity | `c021f17be99c6adfbcba5d38ee0d292c0399d2f5`                         |
| redund_identity | `1925a274ac004373bb5429553bdb55c40e57b124`                         |
| secret_hash     | `1f69c8745f712da03fdd43486ef705fc24f3e34d54cf44d967cf5cd4204c835e` |
| expiry          | 1552263040                                                         |

> Which should compile to the following Bitcoin script bytes:

Reviewer 4: Let's take out the doubts in this message :D 
```suggestion
Which compiles to the following Bitcoin script bytes:
```

While it may seem more intuitive to use a Bitcoin *address* as the identity, the pubkeyhash better fits the definition of identity given in RFC003. Using a Bitcoin address as the identity would require implementations to do a number of cumbersome validation steps such as such as verifying that it is a p2pkh or p2wpkh address, extraction of the pubkeyhash, validation of the network.

When using the pubkeyhash as one of the identities in an RFC003 message body it MUST be encoded as 20 hex encoded string.
This RFC extends the [registry](./registry.md) with the following entry in the identity table:

| Ledger   | Identity Name | JSON Encoding            | Description                                                                                  |
|:----     |:-------       |:-------------            | -------------------------------------------------------------------------------------------- |
| Bitcoin  | `pubkeyhash`  | `hex-encoded-bytes (20)` | The result of applying SHA256 and then RIPEMD160 to a user's SECP256k1 compressed public key |

## Hash Time Lock Contract

### Hash Functions

> The `hash_function` header must be set to one of the following values if Bitcoin is used as a ledger:

Reviewer 2: Yes 🤯 but it's NYI

> The `hash_function` header must be set to one of the following values if Bitcoin is used as a ledger:

Reviewer 2: Also note I was thinking about adding HASH160 as an alternative (so BTC to LTC could be done using that instead). It could be done on Ethereum as well but that would require two contract calls so I'd give it a miss.

    },
    "beta_asset": {...},
    "protocol": "comit-rfc-003",
  },
  "body": {
    "aplha_ledger_refund_identity": "1925a274ac004373bb5429553bdb55c40e57b124",
    "alpha_expiry": 123456789,
    "hash_function": "SHA-256",
    "secret_hash" : "51a488e06e9c69c555b8ad5e2c4629bb3135b96accd1f23451af75e06d3aee9c",
    "beta_ledger_redeem_identity" : "...",
    "beta_expiry" : ...
  },
}
```

And a valid `RESPONSE` could look like:

``` json
{
  "body": {
     "alpha_ledger_redeem_identity": "c021f17be99c6adfbcba5d38ee0d292c0399d2f5",
     "beta_ledger_refund_identity": "...",
  }
}
```

Which would give us the following parameters for the HTLC:

| Parameter       | value                                                              |
|:----------------|--------------------------------------------------------------------|
| redeem_identity | `c021f17be99c6adfbcba5d38ee0d292c0399d2f5`                         |
| redund_identity | `1925a274ac004373bb5429553bdb55c40e57b124`                         |
| secret_hash     | `51a488e06e9c69c555b8ad5e2c4629bb3135b96accd1f23451af75e06d3aee9c` |
| expiry          | 123456789                                                          |

> Both parties should be able to compile the HTLC into this byte sequence:

Reviewer: Should we add what this looks like as a p2wsh address?

Then we would have a complete testcase that allows people to code against.

```

> 6382012088a82051a488e06e9c69c555b8ad5e2c4629bb3135b96accd1f23451af75e06d3aee9c8876a914c021f17be99c6adfbcba5d38ee0d292c0399d2f5670415cd5b07b17576a9141925a274ac004373bb5429553bdb55c40e57b1246888ac

Reviewer: Should we add what this looks like as a p2wsh address?

Then we would have a complete testcase that allows people to code against.
Do we have special cases that might be interesting as test cases?

Reviewer 2: Yes!

Author: If you look at Bitcoin BIPs, then they usually provide a list of tests cases. See [BIP32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki).

Reviewer 2: We at least have one now :)
